python/mlc_llm/compiler_pass/lift_global_buffer_alloc.py

```diff
@@ -1,96 +1,103 @@
 """A compiler pass that lifts TIR-level global allocation to Relax."""
-from typing import Dict, List, Tuple, Optional
+
+from typing import Dict, List, Tuple
 
 import tvm
 from tvm import relax, tir
 from tvm.ir.module import IRModule
+from tvm.relax.analysis import remove_all_unused
 from tvm.relax.expr_functor import PyExprMutator, mutator
 
 
-def LiftTIRGlobalBufferAlloc():
-    @mutator
-    class TIRGlobalAllocRewriter(PyExprMutator):
-        def __init__(self, mod: IRModule):
-            super().__init__(mod)
-            self.mod = mod
-
-        def transform(self) -> IRModule:
-            self.mod = self.builder_.get()
-            for gv, func in self.mod.functions.items():
-                if isinstance(func, relax.Function):
-                    updated_func = self.visit_expr(func)
-                    self.builder_.update_func(gv, updated_func)
-            return self.builder_.get()
-
-        def visit_call_(self, call: relax.Call):
-            call = self.visit_expr_post_order(call)
-            if call.op != tvm.ir.Op.get("relax.call_tir"):
+@tvm.transform.module_pass(opt_level=0, name="LiftTIRGlobalBufferAlloc")
+class LiftTIRGlobalBufferAlloc:  # pylint: disable=too-few-public-methods
+    """A compiler pass that lifts TIR-level global allocation to Relax."""
+
+    def transform_module(
+        self,
+        mod: IRModule,
+        _ctx: tvm.transform.PassContext,
+    ) -> IRModule:
+        """IRModule-level transformation"""
+        return _TIRGlobalAllocRewriter(mod).transform()
+
+
+@mutator
+class _TIRGlobalAllocRewriter(PyExprMutator):  # pylint: disable=abstract-method
+    def __init__(self, mod: IRModule):
+        super().__init__(mod)
+        self.mod = mod
+        self.gv2new_tensor_sinfo: Dict[
+            tvm.ir.GlobalVar, Tuple[tvm.ir.GlobalVar, List[relax.TensorStructInfo], tir.PrimFunc]
+        ] = {}
+
+    def transform(self) -> IRModule:
+        """Entry point of the transformation"""
+        for g_var, func in self.mod.functions_items():
+            # TODO(@eric): This is a temporary hack to get around with two functions for BYOC.
```

Comment:
@Lunderberg, please follow-up.

Reply:
Can you describe what issue is occurring here?  From a quick glance, it looks like this is a bug in the `remove_global_buf_alloc`, that it assumes any `PrimFunc` present in the `IRModule` will be a schedulable PrimFunc.

Reply:
It fails at https://github.com/sunggg/mlc-llm/blob/16c676b461db6f1d3c347aef2ee6f3556dff5b60/python/mlc_llm/compiler_pass/lift_global_buffer_alloc.py#L100